### PR TITLE
Milestone 3 - Friendships v2

### PR DESCRIPTION
**Pull request by jrai0792**
- In this milestone, have added mutual friendship logic.

### CODE REVIEW

File: app/views/users/show.html.erb [hunk 1]
```diff
@@ -1,15 +1,26 @@
...
+      <p>
+        <li><%= user.name %></li>
+      </p>
+    <%end %>
+  </ul>
    <%=form_for(@user, url: user_friendships_path(@user), method: "GET") do |f|%>
-    <%# <%=f.submit "#{Friendship.accepted_request(@user).size} Follower"%>
-    <%# <%=f.submit "#{Friendship.request_confirmed(@user).size} Following"%> 
+    <%=f.submit "Friends #{Friendship.accepted_request(@user).size + Friendship.request_confirmed(@user).size}"%>
+    <%# <%=f.submit "#{Friendship.request_confirmed(@user).size} Invite Sent"  %>
```
Comment (lekegitrepo) [inline app/views/users/show.html.erb]: - _Remove commented code_

File: app/views/users/show.html.erb [hunk 2]
```diff
@@ -1,15 +1,26 @@
...
-    <%unless Invitation.invitation_exists?(current_user.id, @user.id)%>
-      <%= form_for(@user, url: user_invitations_path(@user), method: 'POST') do |f|%>
-        <%= f.submit "Follow"%>  
+      <%unless Friendship.friendship_exists?(current_user.id, @user.id)%>
+        <%unless Invitation.invitation_exists?(current_user.id, @user.id) || Invitation.invitation_exists?(@user.id, current_user.id)%>
+          <%= form_for(@user, url: user_invitations_path(@user), method: 'POST') do |f|%>
+            <%= f.submit "Send Invite"%>  
+          <%end%>
+        <%end%>
       <%end%>
```
Comment (lekegitrepo) [inline app/views/users/show.html.erb]: - _Remove logic from the view_

File: app/views/users/_user.html.erb [hunk 3]
```diff
@@ -1,14 +1,16 @@
...
-        <%unless Invitation.invitation_exists?(current_user.id, user.id) || Invitation.invitation_exists?(user.id, current_user.id)%>
-          <%= form_for(user, url: user_invitations_path(user), method: 'POST') do |f|%>
-          <%= f.submit "Follow"%>  
+        <%unless Friendship.friendship_exists?(current_user.id, user.id)%>
+          <%unless Invitation.invitation_exists?(current_user.id, user.id) || Invitation.invitation_exists?(user.id, current_user.id)%>
+            <%= form_for(user, url: user_invitations_path(user), method: 'POST') do |f|%>
+              <%= f.submit "Send Invite"%>  
+            <%end%>
           <%end%>
         <%end%>
```
Comment (lekegitrepo) [inline app/views/users/_user.html.erb]: - _Remove logic from the view_

File: app/views/users/_user.html.erb [hunk 4]
```diff
@@ -1,14 +1,16 @@
-<%unless current_user== user%>
+  <% unless current_user == user%>
```
Comment (lekegitrepo) [inline app/views/users/_user.html.erb:1]: - _Remove logic from the view_

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -19,7 +19,6 @@
         </div>
         <% if current_user%>
```
Comment (lekegitrepo) [inline app/views/layouts/application.html.erb:20]: - _Remove logic from the view_

File: app/views/friendships/index.html.erb [hunk 6]
```diff
@@ -1,15 +1,15 @@
...
-<%end%>
-</ul>
+<% if @status=="following"%>
+  Friends
+  <ul>
+    <%@friendships.each do |friendship|%>
+    <li><%=friendship.invitee.name%></li>
+    <%end%>
+  </ul>
 <%else%>
```
Comment (lekegitrepo) [inline app/views/friendships/index.html.erb]: - _Remove logic from the view_

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
## Status: Changes Required! :recycle:
### Hi @jrai0792 :clap: :clap:
#### You have done a good job so far :+1: The friendship model works as expected(it creates a friend and inverse friend in the friendship table. However, you need to fix/implement a few requirements before this project can be approved. Check my comments below :point_down:

#### PostsController:
- The index action method should contain code that retrieves data not create one.
```
def index
    @post = Post.new
    timeline_posts
  end
```
`@post = Post.new` should not be in the index action method instead you can create tje `new` action method
```
def new
    @post = Post.new
  end
```

#### Tests:
- _One of the tests is failing :point_down:_ 

Screenshot from 2020-12-17 12-55-41

#### Logic in the view:
- _There should be NO application logic in the views. Everything connected with data processing should be encapsulated in models and used in controllers._

- _View-related logic can be encapsulated in View helpers (app/helpers/\*_helper.rb)_

- _There should be absolutely NO logic in views. >Views are only for displaying information that became available to it from its corresponding controller not generating it. **Check these resources details**_
* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
* [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright

### Optional:
- _You can read about `n+1` and improve your queries if there are `n + 1` queries in index views [Recommended article](https://medium.com/@bretdoucette/n-1-queries-and-how-to-avoid-them-a12f02345be5)._


#### _Reviewed by Temitayo Adeleke_

File: app/views/users/_user.html.erb [hunk 7]
```diff
@@ -1,16 +1,11 @@
...
-        <%unless Invitation.invitation_exists?(current_user.id, user.id) || Invitation.invitation_exists?(user.id, current_user.id)%>
-          <%= form_for(user, url: user_invitations_path(user), method: 'POST') do |f|%>
-          <%= f.submit "Follow"%>  
-          <%end%>
-        <%end%>
-    <%end%>
+    <%= check_user?(current_user, user)%>
+        <%= check_friendship(current_user, user) %>
   </li>
 <%end%>
```
Comment (peterrobert) [inline app/views/users/_user.html.erb]: Kindly remove logic in your views. There should be no logic in your views. Kindly use helpers to remove all the logic in your views.
[Where do I put my code ](http://codefol.io/posts/Where-Do-I-Put-My-Code/)

**Review by peterrobert — CHANGES_REQUESTED** [review]
### CHANGES REQUIRED 🛑 ♻️ 
Hello, @jrai0792  great progress. Good work implementing some of the changes as requested in the previous review. However, there are a few changes that need to be made. Kindly check below ⬇️ 

**logic in views**

- [ ] There are still a few remaining logic in your views. Kindly remember that all logic should be added to the helpers. I will try and name a few places that I have noticed.

1. In your invitations folder under the index.html.erb there is an if statement that is checking for the invitation status.
2.  In your user's folder in the user partial there is an unless condition.

- [ ] Just like stated in the previous review we can not have the new statement in the index method. If you check your posts controller you have this method ⬇️ 
`def index
    @post = Post.new
    timeline_posts
  end`

Instead of that ⬆️ 

create a new method like so ⬇️ 
`def new
  @post = Post.new

end
`
Kindly do those changes and submit your work for another review.

### if you have any question kindly reach me on slack @peterrobert

### HAPPY CODING 💻

**Review by PhillipUg — APPROVED** [review]
# Status: Project Approved! 🥇 
Congratulations on implementing all requirements and getting this project approved. :+1:

You can merge this branch into the development branch when you're ready and go on to your next milestone.
:bulb:  Do not forget to submit your project completion form after merging this branch.

All the best team,

For any assistance, you can contact me on slack as **Phillip Musiime**

**Pull request merged**